File: 0003/c/3.c
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[])
{
    long long num = 0;
    int max_prime = 0, largest_prime = 3;

    if (argc < 2) {
	    puts("Usage: 3 <number>");
	    return EXIT_FAILURE;
    }

    num = strtoll(argv[1], NULL, 10);

    if (!num) {
        fprintf(stderr, "ERROR: %s is not a valid number\n", argv[1]);
        return EXIT_FAILURE;
    }

    printf("Largest prime of %lld: ", num);

    if (num % 2 == 0) max_prime = 2;

    while (num % 2 == 0) num /= 2;

    while (num > 1) {
        while (num % largest_prime == 0) {
            max_prime = largest_prime;
            num /= largest_prime;
        }
        largest_prime += 2;
    }

    printf("%d\n", max_prime);

    return EXIT_SUCCESS;
}

File: 0003/rust/3.rs
use std::env;

fn main() {
    const USAGE: &str = "Usage: ./3.bin <number>";

    let args: Vec<String> = env::args().collect();

    if args.len() == 1 || args.len() > 2 {
        println!("{}", USAGE);
        return;
    }

    let mut number: u64 = match &args[1].trim().parse() {
        Ok(num) => *num,
        Err(_) => {
            println!("{}", USAGE);
            return;
        }
    };

    let input: u64 = number;
    let mut max_prime: u64 = 0;
    let mut largest_prime: u64 = 3;

    if number % 2 == 0 {
        max_prime = 2;
    }

    while number % 2 == 0 {
        number /= 2;
    }

    while number > 1 {
        while number % largest_prime == 0 {
            max_prime = largest_prime;
            number /= largest_prime;
        }
        largest_prime += 2;
    }

    println!("Largest prime of {}: {}", input, max_prime);
}
